### PR TITLE
Adding On-Policy Self-Distillation to Nemo-RL

### PR DESCRIPTION
**Pull request by Hoponga**
# What does this PR do ?

Adding on-policy self-distillation algorithm (https://siyan-zhao.github.io/blog/2026/opsd/) to Nemo-RL, with options to have an EMA-weighted moving teacher, fixed teacher (default), and a teacher sharing the student policy. 

## New Files
1. examples/configs/self_distillation_math.yaml -- config file for self-distillation including generalized JSD as a new loss function
2. examples/run_self_distillation.py -- Driver script for self-distillation
3. nemo_rl/algorithms/self_distillation.py -- Main script for self-distillation training step
4. nemo_rl/data/datasets/response_datasets/openthoughts.py -- added OpenThoughts dataset to Nemo-RL (for self-teacher offline COT reasoning traces)

## Sample Runs 
Running with JSD loss, a fixed teacher, and the Qwen3-4b-Instruct base model, we get the following plots that roughly correspond with the default OPSD implementation: 
image
image

# Issues
List issues that this PR closes ([syntax](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)):

# Usage
Use the run_self_distillation.py script added to examples

```uv run python examples/run_self_distillation.py
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
* ...

## Release Notes

* **New Features**
  * Added self-distillation training workflow with EMA teacher model updates for enhanced model training
  * Added support for OpenThoughts dataset integration
  * Added Jensen-Shannon Divergence (JSD) as a loss option for knowledge distillation

* **Improvements**
  * Improved Ray cluster initialization error handling for better stability

### CODE REVIEW

File: nemo_rl/data/datasets/processed_dataset.py [hunk 1]
```diff
@@ -98,22 +94,11 @@ def encode_single(
...
-            if isinstance(self.task_data_preprocessors, dict):
-                task_name = entry["task_name"]
-                if task_name in self.task_data_preprocessors:
-                    task_data_preprocessor = self.task_data_preprocessors[task_name]
-            else:
-                task_data_preprocessor = self.task_data_preprocessors
 
-        if task_data_preprocessor is not None:
-            entry = task_data_preprocessor(entry)
+        #print(self.dataset)
```
Comment (yuki-97) [inline nemo_rl/data/datasets/processed_dataset.py:103]: could you remove the debug codes? and other places' debug codes as well?

File: nemo_rl/algorithms/self_distillation.py [hunk 2]
```diff
@@ -0,0 +1,1550 @@
...
+        try:
+            return [default if v is None else str(v) for v in value.tolist()]
+        except Exception:
+            return [default for _ in range(batch_size)]
+    # Scalar fallback: repeat for batch
+    return [default if value is None else str(value) for _ in range(batch_size)]
+
+
+
+def _build_cot_gt_texts(
```
Comment (yuki-97) [inline nemo_rl/algorithms/self_distillation.py:190]: it looks like this function is only using cot and ground truth which seems unrelated to the runtime.

shall we move this process to `OpenThoughtsDataset` 's `format_data`, and then we could just record `cot_text` in each sample and don't need to pass `deepseek_reasoning`, `ground_truth_solution` and `deepseek_solution` in many places?

File: nemo_rl/data/datasets/response_datasets/openthoughts.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+        return {
+            "messages": [
+                {"role": "user", "content": problem},
+                {"role": "assistant", "content": assistant_answer},
+            ],
+            "task_name": self.task_name,
+            "problem": problem,
+            "ground_truth_solution": ground_truth_solution,
+            "deepseek_reasoning": deepseek_reasoning,
```
Comment (yuki-97) [inline nemo_rl/data/datasets/response_datasets/openthoughts.py:63]: wondering whether `deepseek_solution` is missing here? I saw many `deepseek_solution` at other places.

File: .gitignore [hunk 4]
```diff
@@ -17,6 +17,9 @@ dist/
 *.vscode/
 release_run*
 ckpts/
+*.so
```
Comment (yuki-97) [inline .gitignore:21]: could you revert the changes in this file and the temp files `run_sd.sh` and `sd_log`?

File: nemo_rl/__init__.py [hunk 5]
```diff
@@ -36,6 +36,12 @@
 if megatron_path.exists() and str(megatron_path) not in sys.path:
     sys.path.append(str(megatron_path))
 
+automodel_path = (
```
Comment (yuki-97) [inline nemo_rl/__init__.py:39]: just curious why will you need this? ideally `uv run python examples/run_self_distillation.py` could run without this.

**Review by yuki-97 — COMMENTED** [review]
hi @Hoponga , thanks for adding this, nice work!

I reviewed partly and left some comments. and I think you may need to do a rebase since your branch is a bit outdated from main (50+ commits behind).

**Comment by Hoponga** [comment]
Hi @terrykong @yuki-97 , thanks for your reviews and comments. Subclassing the PolicyWorker's for the additional ema-updating logic makes sense -- let me know when the implementation is ready and I'll update my code accordingly! And I'll also see if I can get some more comprehensive results for the self-distillation evals that correspond with the original OPSD paper.

File: nemo_rl/algorithms/loss/loss_functions.py [hunk 6]
```diff
@@ -926,6 +1118,15 @@ def __call__(
             per_token_kl = student_probs * (
                 student_topk_logprobs - teacher_topk_logprobs
             )
+        elif self.kl_type == "jsd":
```
Comment (terrykong) [inline nemo_rl/algorithms/loss/loss_functions.py]: [`loss_functions.py:1121`](https://github.com/NVIDIA-NeMo/RL/blob/481065a198233515847b268d453379e34b363a6f/nemo_rl/algorithms/loss/loss_functions.py#L1121)

**TEST**: The new `jsd` KL type branch is not covered by any test. The existing `test_distillation_loss_different_settings` parametrizes over `["forward", "reverse", "mixed"]` only. Please add test coverage for:
- Numerical correctness vs. a reference JSD implementation
- The `zero_outside_topk` × `jsd` combo (exercises the `P_rest * 0.5 * log(2)` tail correction at line 1111)
- Non-negativity property of JSD
- Gradient flow through the `logaddexp` path

File: nemo_rl/algorithms/loss/loss_functions.py [hunk 7]
```diff
@@ -904,6 +924,173 @@ def __call__(
...
+        # CP support: get CP group and size
+        cp_group = context_parallel_group
+        cp_size = 1 if cp_group is None else torch.distributed.get_world_size(cp_group)
+
+        # Ensure float32 for stability (match other losses)
+
+        # next token logits is local to one GPU -- post-sharded tensor of size [B, S_local, V_local]
+        # S_local is depenent on context parallelism 
+        # V_local is dependent on vocab/tensor parallelism
+        next_token_logits = next_token_logits.to(torch.float32)
```
Comment (terrykong) [inline nemo_rl/algorithms/loss/loss_functions.py]: [`loss_functions.py:941`](https://github.com/NVIDIA-NeMo/RL/blob/481065a198233515847b268d453379e34b363a6f/nemo_rl/algorithms/loss/loss_functions.py#L941)

**BUG (Critical)**: The new code block added inside `DistillationLossFn.__call__` references variables that don't exist in the function scope: `next_token_logits`, `context_parallel_group`, `vocab_parallel_group`, `vocab_parallel_rank`, plus unimported functions (`gather_logits_at_global_indices`, `ChunkedDistributedGatherLogprob`, `ChunkedDistributedEntropy`, `allgather_cp_sharded_tensor`, `_get_tokens_on_this_cp_rank`). This will raise `NameError` on the first training step.

This block looks like a partial copy-paste from worker-side code that was never properly integrated with the loss function interface. Either remove this block so `__call__` starts at the original `student_probs = student_topk_logprobs.exp()` line, or move the logic into the worker where these variables exist.

File: nemo_rl/algorithms/distillation_worker_extension.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+import ray
+import torch
+
+from nemo_rl.models.policy.utils import get_runtime_env_for_policy_worker
+from nemo_rl.models.policy.workers.dtensor_policy_worker_v2 import (
+    DTensorPolicyWorkerV2Impl,
+)
+
+
+class DistillationWorkerV2Impl(DTensorPolicyWorkerV2Impl):
```
Comment (terrykong) [inline nemo_rl/algorithms/distillation_worker_extension.py]: [`distillation_worker_extension.py:30`](https://github.com/NVIDIA-NeMo/RL/blob/481065a198233515847b268d453379e34b363a6f/nemo_rl/algorithms/distillation_worker_extension.py#L30)

**BUG**: `DistillationWorkerV2Impl` only extends `DTensorPolicyWorkerV2Impl`, but the setup code unconditionally passes this extension for the student policy (line 766). If a user selects the Megatron backend (supported in the config), the extension's DTensor-based methods won't work correctly under Megatron sharding. Either gate the extension to DTensor-only, or add a parallel `DistillationMegatronWorkerImpl` that extends `MegatronPolicyWorkerImpl`.

File: nemo_rl/algorithms/self_distillation.py [hunk 9]
```diff
@@ -0,0 +1,1574 @@
...
+    """Update teacher parameters using exponential moving average (EMA) of student parameters.
+
+    Formula: teacher_param = ema_decay * teacher_param + (1 - ema_decay) * student_param
+    Uses worker extension methods (get_model_state_dict, update_model_with_ema)
+    via run_all_workers_single_data instead of modifying the base worker classes.
+    """
+    if ema_decay >= 1.0:
+        return
+
+    student_states = student_policy.run_all_workers_single_data("get_model_state_dict")
```
Comment (terrykong) [inline nemo_rl/algorithms/self_distillation.py]: [`self_distillation.py:453`](https://github.com/NVIDIA-NeMo/RL/blob/481065a198233515847b268d453379e34b363a6f/nemo_rl/algorithms/self_distillation.py#L453)

**MEMORY**: `_update_ema_teacher` calls `ray.get()` on ALL worker state dicts, materializing the full model weights on the driver, then sends each dict back out via `.remote()`. For a multi-billion-parameter model this round-trips many GB through the driver every training step and will likely OOM. Consider using Ray ObjectRefs (pass futures instead of materialized dicts) or direct worker-to-worker communication.
